Make the pane resizable on mobile

FivEawE opened the pull request:
Hello. Since I saw #9 was open, I thought I would give it a go and try to solve it. This is an initial implementation, I would like to get some points clarified before finishing it. I tried using a similar approach as the one used for horizontal resizing.
First of all, I made the div responsible for resizing 20px tall. It is not very pretty but it correctly registers touch movements unlike its smaller counterpart. I was thinking of making it as small as the one for horizontal resizing and making an invisible touch area around it, but I am not sure if it would collide with the Result and Output buttons. I can check it out if you would like to go this path.
On the same topic, I was thinking of putting a three grey dots (or something similar) inside the div as an indicator of the possibility to resize the panes, since hovering on the phone is triggered on click and it might not be as obvious as on the desktop.
I was also thinking about the indicator which is displayed on hover. I was thinking of enabling the effect not only on hover, but also during the process of resizing. If you move the mouse too quickly and the cursor is not hovering over the div, the effect is lost. It is even worse on phone.
For "normalizing" the measurements of mouse/touch events I used a couple of refs to get the heights of the surrounding non-resizeable elements. It was possible to hardcode the measurements from CSS, but I thought this was a better approach in case the layout changes in the future. Also the calculations could be put in a memo instead of recalculating them on every onMouse/onTouch movement, but I do not think there would be a significant performance gain.
These are the points I would like clarification on. I am open for discussion and other implementation changes or suggestions.

amoutonbrady commented:
Hey! Thanks for taking on that one.

The issue was kind of mistitled in the sense that I mostly wanted to be able to allow vertical resizing when the panes are on top of each other (which mostly happen on mobile, but not only).

What you started working on is much deeper than I had in mind. I'm still trying to figure out how mobile playground could work. The svelte REPL has two kind of tabs, with a switch to toggle between code and preview. I wonder if that's the play here...

I feel like making the panes resizable on mobile is a lot of work for not much benefits on the end user side of things.

Other than that everything you mentioned makes total sense and seem reasonable:

- bigger resizing separator with an icon hint
- resizing hint
- auto calculation of the panes to resize

From all of this I'm thinking, let's make it work on vertical, primarily for mouse user (which would then be able to be used on the tutorial part of the website), then we can think about what the plan is for actual mobile device users.

For what it's worth, we have a query params wrongly named `isHorizontal` which allow you to see the playground in vertical mode (where the resizing is not possible) like so : https://playground.solidjs.com?isHorizontal

FivEawE commented:
Hello, thanks for clarifying the issue. I adjusted resizing according to the provided information.

For the design of the resizers I picked already used colors. On hover and during the resizing state I used the main brand color for the resizer. I also tried using couple of shades of grey to see how it would look, but in my opinion, the brand color worked the best. You can see the design in the screenshots below.

![1](https://user-images.githubusercontent.com/25206541/127767922-e60ea438-2bf4-4647-9af0-511d72090a80.png)
![2](https://user-images.githubusercontent.com/25206541/127767923-de42a1f4-c59a-4a96-a21d-1ea68e4df3e6.png)
![3](https://user-images.githubusercontent.com/25206541/127767924-07abafcf-6d12-4929-be04-cd43b0390af5.png)
![4](https://user-images.githubusercontent.com/25206541/127767925-f961803a-0389-447a-8760-4baf147db403.png)

I have made the resizer slightly smaller than in the original PR so it would not occupy that much space. Since attaching a touch listener was easy, I decided to implement it as well. It can be kept as a temporary solution. To my surprise it works fine even on phones with larger screens.

I separated the resizer into its own component. I decided to extend the div interface in case somebody would like to change some of the aspects of the resizer in the future without them changing the behavior of the component itself. Even though the resizer has its classes encapsulated, I decided that it would be possible to extend the classes with custom ones as well. For instance, it makes it easier to change resizer's position in the grid.

While in the resizing state, I decided to overlay the whole screen with a div with fixed position and z-index of 50. By doing so we can keep displaying the resizing cursor without it switching to pointer or text variants when hovering over the elements. It also prevents text selection while resizing. Because of this, I removed the pointer-events-none from Preview since the overlaying div makes it impossible to trigger any pointer event in the iframe.

I also adjusted the percentage calculation because the original one had a flaw. For instance, if the left pane had the percentage set to 0.6 and you moved your mouse too quickly, the percentage calculation could result in 0.4 which would not be accepted as valid, yet the pane still would have 0.1 percentage left to move.

From my perspective, I think the implementation is complete unless you would like me to change something. I can merge/rebase on (depending on what you prefer) the current master for easier integration.

amoutonbrady commented:
That's awesome, really.

Thanks a lot for the details and the quick adjustment. This will be a very very welcomed change by a lot of people :)

Much appreciate it the detailed writeup!
